src/ripple/basics/SlabAllocator.h

```diff
@@ -0,0 +1,235 @@
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <mutex>
+
+namespace ripple {
+
+template <typename Type, std::size_t ExtraSize = 0>
+class SlabAllocator
+{
+public:
+    struct Stats
+    {
+        // The name of the allocator
+        std::string name;
+
+        // The size of an individual item
+        std::size_t size = 0;
+
+        // The number of items the allocator can support
+        std::size_t count = 0;
+
+        // How many allocation requests were made to the allocator
+        std::uint64_t alloc_count = 0;
+
+        // How many of allocations have been serviced by the allocator
+        std::uint64_t alloc_fast_count = 0;
+
+        // How many of deallocations have been serviced by the allocator
+        std::uint64_t dealloc_fast_count = 0;
+    };
+
+private:
+    // The name of this allocator, used for debugging purposes:
+    std::string const name_;
+
+    // The number of items to allocate:
+    std::size_t const count_;
+
+    // The underlying memory block:
+    std::uint8_t* p_;
+
+    // A linked list of free buffers:
+    std::uint8_t* l_ = nullptr;
+
+    // A mutex to protect the linked list:
+    std::mutex mutable m_;
+
+    // How many allocations we've serviced:
+    std::atomic<std::uint64_t> alloc_count_ = 0;
+
+    // How many of those were from our internal buffer:
+    std::atomic<std::uint64_t> alloc_fast_count_ = 0;
+
+    // How many of those went to our internal buffer:
+    std::atomic<std::uint64_t> dealloc_fast_count_ = 0;
+
+    // Scrubs the memory range [ptr, ptr+Size)
+    std::uint8_t*
+    scrub(std::uint8_t* ptr, [[maybe_unused]] std::uint8_t value)
+    {
+        assert(ptr != nullptr);
+
+#ifdef SLAB_SCRUB_MEMORY
+        std::memset(ptr, value, Size);
+#endif
+
+        return ptr;
+    }
+
+public:
+    // We may need to adjust the size of an individual item to account for
+    // alignment issues:
+    constexpr static std::size_t const size = []() constexpr {
+        std::size_t ret = sizeof(Type) + ExtraSize;
+
+        if (auto const mod = (sizeof(Type) + ExtraSize) % alignof(Type))
+            ret += (alignof(Type) - mod);
+
+        return ret;
+    }();
+
+    /** Constructs a slab allocator able to allocate up to count items
+
+        @param count the number of items the slab allocator can allocate; note
+                     that a count of 0 is valid and means that the allocator
+                     is, effectively, disabled. This can be very useful in some
+                     contexts (e.g. when mimimal memory usage is needed) and
+                     allows for graceful failure.
+     */
+    SlabAllocator(
+        std::size_t count,
+        std::string name = "Slab: " + beast::type_name<Type>())
+        : name_(name), count_(count)
+    {
+        if (count != 0)
+        {
+            p_ = reinterpret_cast<std::uint8_t*>(
+                std::aligned_alloc(alignof(Type), size * count_));
+            assert(p_ != nullptr);
```

Comment:
Major snag:  `aligned_alloc` always returns `nullptr` using very latest macOS and tools.

Reply:
I replaced `alignof(Type)` with `(alignof(Type) >= 8 ? alignof(Type) : 8)` (everywhere) and that got `aligned_alloc` to return non-null pointers.  But I then got tons of unittest failures.